### ISSUE TITLE
TypeError when run with docker in Ubuntu 14.04

### ISSUE THREAD

**Issue by moon-musick** [metacloud/molecule#548]
Hi, 

I am a new molecule user. I am trying to run molecule with Docker. It seems I can get it to work properly with Vagrant without any issues.

I get the following traceback when running `molecule test` after `molecule init --driver docker` in an existing role directory:

```
--> Destroying instances ...
--> Checking playbook's syntax ...

playbook: playbook.yml
--> Creating instances ...
Traceback (most recent call last):
  File "/usr/local/bin/molecule", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/molecule/cli.py", line 41, in main
    cli(obj={})
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/molecule/command/test.py", line 99, in test
    util.sysexit(t.execute()[0])
  File "/usr/local/lib/python2.7/dist-packages/molecule/command/test.py", line 46, in execute
    status, output = c.execute(exit=False)
  File "/usr/local/lib/python2.7/dist-packages/molecule/command/create.py", line 44, in execute
    self.molecule.driver.up(no_provision=True)
  File "/usr/local/lib/python2.7/dist-packages/molecule/driver/dockerdriver.py", line 117, in up
    self._build_ansible_compatible_image()
  File "/usr/local/lib/python2.7/dist-packages/molecule/driver/dockerdriver.py", line 226, in _build_ansible_compatible_image
    for tag in image.get('RepoTags')]
TypeError: 'NoneType' object is not iterable

```

The software version is as follows:

```
Linux e450 4.2.0-42-generic #49~14.04.1-Ubuntu
Docker version 1.12.1, build 23cf638

docker-compose==1.7.0
docker-py==1.10.3
docker-pycreds==0.2.1
dockerpty==0.4.1
molecule==1.12.4
testinfra==1.4.2
```

The user I run molecule with is in `docker` group, able to run docker stuff without `sudo`.

Is there anything wrong with my setup, or perhaps I misunderstand something about how molecule works?

**Comment by retr0h** [comment]
I have seen this [once before](https://github.com/metacloud/molecule/issues/489).  I cannot tell you why it happens, but deleting your docker data dir should correct the issue.  Might have been an issue left behind when upgrading from an older version of docker.

**Comment by moon-musick** [comment]
What do you mean by Docker data dir? Is it `/var/lib/docker`?

**Comment by retr0h** [comment]
I'm not entirely sure where linux stores this directory.  Also, deleting the directory would require some thought.  It would remove all docker images you have downloaded or used.

You could also try removing all images through the CLI as well (`docker rmi $(docker images -q)`).

**Comment by vterdunov** [comment]
I got the same problem on Linux (Xubuntu 16.04). I've fixed it by delete docker data (`/var/lib/docker`). And reinstalled docker.

**Comment by retr0h** [comment]
Did you have old images in there?  I'm curious why the Docker API returns images w/o a `RepoTags` attribute.

**Comment by moon-musick** [comment]
It's 6.5 GB of locally built images in my case, and build process for some of those is quite CPU-intensive. I prefer not to delete all this if I don't really have to.

Is it perhaps possible to isolate this API component in some test script?

**Comment by retr0h** [comment]
It's not a molecule problem, it's a Docker API issue.  It's not returning necessary data back.

**Comment by moon-musick** [comment]
I guess it's either one of molecule's library dependencies (`docker-py`?) or docker engine itself which is at fault. Am I able to do something on my part which would help debug this issue?

**Comment by retr0h** [comment]
You could try to see which image is reporting no `RepoTag`.

```
import docker
_docker = docker.Client(version='auto')

created_containers = _docker.containers(all=True)

for image in _docker.images():
    if not image.get('RepoTags'):
        print image['Id']
```

Then run the returned id through `docker images --no-trunc  | grep $id`

**Comment by moon-musick** [comment]
It seems it was an image without a tag indeed, a custom Nginx build. I retagged the next build with the same tags. Your script helped a lot.

If I understand correctly, either molecule or some library dependency raises an uncaught exception if there is an image without a tag. If this works like that on purpose (e.g. because of a policy to fail early on wrong input / environment, if tags are somehow a crucial factor), perhaps a customized error message explaining the issue might be a good idea?

**Comment by retr0h** [comment]
I'd like to understand how an image gets installed without a tag.  Is that a normal or out of the ordinary?  I've only run across this a couple of times, and it feels like the user created the image incorrectly, but I would like to know how one can upload an untagged image to trigger this bug.

**Comment by retr0h** [comment]
I'd like to know how a system would get into this state.  I have a few images listed as none tag, yet I cannot trigger this bug.

The `images()` call returns only 5 images and ignores the others.

```
[jodewey:~/git/molecule/test/scenarios/group_host_vars] bug-548(+6/-1)+ ± docker images -a
REPOSITORY                TAG                 IMAGE ID            CREATED             SIZE
molecule_local/ubuntu     latest              7ead251fdcb1        7 days ago          200.5 MB
<none>                    <none>              f69218ebf65e        7 days ago          200.5 MB
molecule_local/opensuse   latest              48a8f621e6f6        8 days ago          232.3 MB
<none>                    <none>              ac0568d0a60a        8 days ago          97.78 MB
<none>                    <none>              d3d5b779ca94        8 days ago          97.78 MB
ubuntu                    latest              c73a085dc378        2 weeks ago         127.1 MB
alpine                    latest              ee4603260daa        2 weeks ago         4.803 MB
opensuse                  latest              642bc8a45961        12 weeks ago        97.78 MB
```

**Comment by retr0h** [comment]
@moon-musick I would be interested to know if the [bug-548](https://github.com/metacloud/molecule/tree/bug-548) branch corrects the problem for you.  Is that something you could verify?

**Comment by moon-musick** [comment]
I already deleted the apparently faulty image to check if it would help and it seems molecule runs docker containers properly now.

I tried creating some images without tags to test if I can still reproduce the issue, but I can't, even if I delete images and containers created by molecule, the `.molecule` folder in my role folder and re-init molecule.

The container in question had the `REPOSITORY` column (in `docker images` output) set to `nginx`, but the `TAG` column was `<none>`. I didn't run `docker inspect` on it, so I can't say what the `RepoTags` value was really set to. At the moment I have no idea how to re-create an image with metadata like that.
